src/transformStateWithClones.js

```diff
@@ -7,7 +7,46 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state }; // Create an initial copy of the state
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = addProperties(currentState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        currentState = removeProperties(currentState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        currentState = clearProperties(currentState);
+        break;
+    }
+    stateHistory.push(currentState);
```

Comment:
The current implementation pushes the `currentState` to `stateHistory` inside the `switch` block. According to the additional prompt rules, it is better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed. Consider moving this line outside of the `switch` block, just before the end of the loop.